flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java

```diff
@@ -557,11 +567,84 @@ public List<JavaScriptInvocation> dumpPendingJavaScriptInvocations() {
             return Collections.emptyList();
         }
 
-        List<JavaScriptInvocation> currentList = pendingJsInvocations;
+        List<JavaScriptInvocation> readyToSend = getPendingJavaScriptInvocations()
+                .stream()
+                .filter(invocation -> invocation.getOwner().map(StateNode::isVisible).orElse(true))
+                .collect(Collectors.toList());
+
+        pendingJsInvocationDetachListeners.entrySet().stream()
+                .flatMap(invocationListeners -> readyToSend.stream()
+                        .filter(invocation -> invocation.getOwner()
+                                .orElseGet(() -> ui.getInternals().getStateTree().getRootNode())
+                                .getId() == invocationListeners.getKey().getId()))
+                .forEach(invocation -> pendingJsInvocationDetachListeners.values()
+                        .forEach(listener -> listener.removePendingInvocation(invocation)));
```

Comment:
It seems a bit complicated to me. Also, it should call `listener.onInvocationCompleted()` instead of `listener.removePendingInvocation()` otherwise the invocation will not be removed from the `JavaScriptInvocationDetachListener`

And as commented before, method `onInvocationCompleted` should be renamed with something more related to the context where it is applied in this PR

```suggestion
        StateNode rootNode = ui.getInternals().getStateTree().getRootNode();
        readyToSend.forEach(i -> {
            JavaScriptInvocationDetachListener listener = pendingJsInvocationDetachListeners.get(i.getOwner().orElse(rootNode));
            if (listener != null) {
                listener.onInvocationCompleted(i);
            }
        });
```